File: python/proverb/proverb.py
def proverb(*lst, qualifier):
    """
    Generate a proverbial saying based on a list of items and an optional qualifier.

    This function generates a proverbial saying based on a list of items, where each item
    represents a cause of the next item's loss, and an optional `qualifier` can be included
    to modify the first item in the list.

    Args:
        *lst: A variable number of items (strings) representing causes of loss.
        qualifier (str, optional): An optional qualifier to modify the first item.

    Returns:
        list: A list of strings forming the proverbial saying.

    Example:
        >>> proverb("nail", "shoe", "horse", "rider", "message", "battle", "kingdom")
        [
            "For want of a nail the shoe was lost.",
            "For want of a shoe the horse was lost.",
            "For want of a horse the rider was lost.",
            "For want of a rider the message was lost.",
            "For want of a message the battle was lost.",
            "For want of a battle the kingdom was lost."
        ]
    """
    result = []
    if len(lst) > 1:
        for counter, element in enumerate(lst[:-1]):
            result.append(f"For want of a {element} the {lst[counter + 1]} was lost.")
    if len(lst) != 0 and qualifier != None:
        result.append(f"And all for the want of a {qualifier} {lst[0]}.")
    elif len(lst) != 0:
        result.append(f"And all for the want of a {lst[0]}.")
    return result

File: javascript/proverb/proverb.js
export const proverb = (...items) => {
  const qualifier =
    typeof items[items.length - 1] === "object"
      ? items.pop().qualifier + " "
      : "";
  const lines = [];

  for (let i = 0; i < items.length - 1; i++) {
    lines.push(`For want of a ${items[i]} the ${items[i + 1]} was lost.`);
  }

  if (items.length > 0) {
    lines.push(`And all for the want of a ${qualifier}${items[0]}.`);
  }

  return lines.join("\n");
};
